src/librustc_typeck/astconv.rs

```diff
@@ -413,9 +413,20 @@ fn create_substs_for_ast_path<'tcx>(
     let mut type_substs = if param_mode == PathParamMode::Optional &&
                              types_provided.is_empty() {
         let mut substs = region_substs.clone();
+
         ty_param_defs
             .iter()
-            .map(|p| this.ty_infer(Some(p.clone()), Some(&mut substs), Some(TypeSpace), span))
+            .map(|p| {
```

Comment:
This code *rather* needs some refactoring (at least, split it into another method).